### ISSUE TITLE
Download and install the contrib modules listed in the case study

### ISSUE THREAD

**Issue by lhridley** [savaslabs/autism-case-study#5]
Depends on Issue #2, #3, and #4
* Download the contrib modules indicated
* Write a hook_update_n method in "acs_master" to enable any core and contrib modules not enabled in the seed database that should be enabled based on the case study instructions.
* Submit a PR for review

References:
* https://www.drupal.org/node/2302845
* http://www.metaltoad.com/blog/how-write-drupal-update-scripts

**Comment by lrowang** [comment]
Hi @lhridley ,

1. All the contrib modules downloaded fine except the `features_extra` module. When I tried to `drush pm-enable features_extra -y` in the VM, I got stuck on an infinite loop. I did a little research and it seems that this is a known [Drush bug](https://github.com/drush-ops/drush/issues/5), but I couldn't find a solution. Is there a way of resolving this via Drush or do I just have to enable `features_extra` via the UI?

2. When I enabled `date` module, I got this warning, "The Date API requires that you set up the site timezone and first day of week settings and the date format settings to function correctly." I changed the settings, and since this is a configuration and not a change in code, I know it doesn't get captured by git. BUT since I don't have any kind of hook written that interfaces with the DB (yet), does this mean these changes will be lost when I power down the VM and spin it back up next time?

3. I did my best to write what I think is 2 possible `hook_update_N` methods for `acs_master`. Though I'm pretty sure they're not right... There was a lot of info out there and I'm a little confused by it all!

**Comment by lrowang** [comment]
Hi @lhridley ,

I've pushed an update to the PR. I do have a lot of questions!

1. First, I arranged the functions within the `.install` file in an order that made sense to me (things needing least change to stuff that I think will likely get edited more often) as I couldn't find any best practice suggestions on how to structure the code within `.install`.

2. I designated each for loop function as an "internal" function with the `_` prefix in the name. Was this necessary? 

3. Related to the for loop, I noticed the HPTN modules were enabled differently in their `.install` file. Are there multiple ways to enable modules (and is one preferred over others?), or is the syntax different because HPTN is D8?

4. When reading about the install and uninstall hooks, there were a lot of references to hook_schema` and DB tables -- how does that relate to `hook_update_N`?

5. Lastly, I'm not sure how to "test" if my code changes are working. I spun up the VM and loaded the site and went to `admin/modules` and I can see that the Core Aggregator module is not enabled even though it should be from the `.install` file. Am I missing another step?

**Comment by lhridley** [comment]
1.  Typically there is an order:

- hook_install()
- hook_uninstall()
- private helper methods
- hook_update_n() methods

The order of private helper methods vs. hook_update_n is a subject for debate. I like to include them before the hook_update_n methods so that anyone who comes in after me to add a hook_upate_n() sees that there are helper methods available -- increases the likelihood that they will be used as designed.

For the hook_update_n() methods, while there is no rule, I like to put them in reverse numeric order.  This puts the newer methods first, so anyone reviewing them doesn't have to scroll to the end of the file to find them.

2.  No necessary, but good practice.   Since Drupal 7 is written as a functional application, there is no "public" vs. "private" methods.  It's standardly accepted in the community that method names prefixed with an underscore are considered "private" and should not be called by functions outside the module to which these methods belong (they are not to be considered part of the module's API).

3.  Yes, this is a difference between D7 and D8

4.  hook_schema is another method that is part of the .install files.  You include a hook_schema method when you are creating a database table that is part of the module's functionality.  Drupal uses hook_schema to subsequently create the table, and if a module has a hook_schema method defined and it is removed from a project using pm-uninstall (or the internal drupal method for uninstalling a module), the table will be deleted as part of the uninstall process.

5.  The order in which modules are enabled can be important -- if the module has dependencies that are missing, it will not be enabled.  Also, you may not have the machine name of the module specified correctly.

**Comment by lrowang** [comment]
Hi @lhridley ,

PR #15 updated! I merged the updated `develop` branch with Docker into my feature branch and continued with some of the changes we had discussed previously last week:

1. One major formatting change I made was to create a separate array listing all the modules, then pass that array to both `acs_master_install` and `acs_master_uninstall`. My reasoning is I only have to create the array once (making it easier to add another module later if needed), and also I don't have to recreate it all over again inside `acs_master_unstinall` (I can just pass it and reverse it for the disabling). Please let me know if this method is acceptable!

2.  I went through all the contrib module's `.info` files to check their dependencies. I believe they are now listed in the appropriate order. 

Questions:
1. Could we walk through how to implement  my `hook_install` and `hook_update_#` (or even how to check they are working correctly)? I think this is done by shelling in and executing drush commands? 

2. And on a related note, since there is no longer a `vagrantfile` and startup shell script, where does ACS Master get enabled for the first time?

3. This is a larger project workflow question: if I merge a PR/branch, am I responsible for deleting it on the main project repo? Or is it best practice to leave that to the PM?

**Comment by lrowang** [comment]
Hi @lhridley ,

I generated a new PR #22 in the main SavasLabs repo so Travis can build on it. The update includes all the changes we discussed today: 

- helper array function
- hook_update_# from 7500 to 7100
- tidying comment formats

Thanks!
Ro

**Comment by lrowang** [comment]
Hi @lhridley , 

I pushed updates to PR #22. Changes include:
- Rebased commits from the Wodby container updates (#24)
- Error handling and enable_module fixes in the `acs_master.install` file
- Makefile updates you requested in slack

After making all those changes, I ran `make clean`, and `make install` again to see how it would execute. 
- Good news: site is up and I didn't get any of the big warning chunks like before!
- Bad news: I didn't see the confirmation message "Enabled <array of modules>" that I wrote in the hook_install. And when I checked '/admin/modules in the UI, I could see that none of them were enabled! I would have expected hook_install to run again since I blew away the environment by `make clean` -- or is that incorrect?

Ro
